experimental/uefi/src/main.rs

```diff
@@ -51,14 +62,57 @@ fn _start(_handle: Handle, mut system_table: SystemTable<Boot>) -> Status {
         .reset(ResetType::Shutdown, status, None);
 }
 
-fn main(_handle: Handle, system_table: &mut SystemTable<Boot>) -> Status {
+fn main(handle: Handle, system_table: &mut SystemTable<Boot>) -> Status {
     use core::fmt::Write;
 
-    let status = writeln!(system_table.stdout(), "Hello World!");
+    writeln!(system_table.stdout(), "Hello World!").unwrap();
+
+    serial_echo(handle, system_table.boot_services()).unwrap();
+}
+
+fn echo_loop(serial: &mut uefi::proto::console::serial::Serial) -> Result<!, uefi::Error<()>> {
+    let mut buf: [u8; 1024] = [0; 1024];
+
+    loop {
+        // read() returns Ok if it managed to fill the whole buffer, or the error will contain
+        // the number of bytes read. The only error we're fine with is TIMEOUT, as we can simply
+        // retry that (and we'll keep getting TIMEOUTs when nobody is talking to us). In case of
+        // any other error, bail out.
+        let len = serial.read(&mut buf).map(|_| buf.len()).or_else(|err| {
+            if err.status() == Status::TIMEOUT {
```

Comment:
Does it mean that if no one is talking to us, this will be a busy loop? Or is there an actual timeout that it waits for incoming data before hitting this?

Reply:
Yes, this is a busy loop, for now.

By default, UEFI gives us (a) an one-second timeout and (b) a buffer size of one byte. Both of these can be changed (and we definitely need to make the buffer larger to not lose any data in any real scenario); in the future we'd likely need to do some kind of flow control to ensure that we're not sending data while we haven't processed what's in the buffer yet.